# Inferential Statistics:-
#### It is used together ,some information about the population from the sample of data collected.There are many tests which are used in inferential statistics,depending on the objective and the data.
#### Objective 1:-To find out ,is there a significant difference in the marks of students between section A and section B.For any objective we formulate a hypothesis.
#### Null Hypothesis(Ho)=There is no significant difference in the marks between students of section A and B.
#### Alternate Hypothesis(Ha)=There is a significant difference in the marks of Section A and B.
### The test to be used:-
#### Step1- Check the normality for the variable wingA and WingB separately.
#### Step2- If both the variable have normal distribution then we use a parametric test known as 'Independent sample t-Test'.If the variables don't have the normal distribution then we use a non-paramtric test known as 'Mann Whitney U Test'.
### How to do Independent Sample t-Test in python:-
#### Import the library and import the dataset into python.
#### To do the Independent Sample t-Test in python,the function used is ttest_ind.
### How to import a specific function from the library:-
#### TO import the function 'ttest_ind' from the python library 'scipy.stats'

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('D:\Section.csv')

#### There are two variables to be used for dataframe.The independent sample t_test will give a t value and this t value has to be compared with the t critical value.

#### t_statistics,p_value=stats.ttest_ind(df['wingA'],df['wingB']) #to calculate t_statistics
##### type t_statistics and run the code.

#### The value from the python is the t calculated.This value has to be compared wih the critical value of t.This critical value will be obtained from the t table. Use the column for 5% significance level.Degreee of freedom =(n-2) where n is the sample size.
#### Get the Crtitical value from the statistical table.

## 1.) If the calculated t from python is less than the critical value from the table then you accept the Null Hypothesis and conclude there is no difference in the performance of the students between the two wings.
## 2.) If the calculated t from python is greater than the critical value from the table then you accept the Alternate Hypothesis and conclude there is a difference in the performance of the students between the two wings.

In [3]:
from scipy.stats import ttest_ind

In [4]:
df.head(20)

wingA  wingB
0   10.70   7.20
1    9.89   6.68
2   11.83   9.29
3    9.04   8.95
4    9.37   6.61
5   11.68   8.53
6    8.36   8.92
7    9.76   7.95
8   13.67   7.57
9    8.96   6.38
10   9.51   8.89
11  10.85  10.03
12  10.57   9.30
13  11.06   5.28
14   8.91   9.23
15  11.79   9.25
16  10.59   8.44
17   9.13   6.57
18  12.37  10.61
19   9.91   6.77

In [5]:
freq_table=pd.crosstab(df['wingA'],'count')  #crosstab is the function in pandas library.
pd.set_option('display.max_rows',None) #To display the full freq table.
freq_table

col_0  count
wingA       
8.36       1
8.91       1
8.96       1
9.04       1
9.13       1
9.37       1
9.51       1
9.76       1
9.89       1
9.91       1
10.57      1
10.59      1
10.70      1
10.85      1
11.06      1
11.68      1
11.79      1
11.83      1
12.37      1
13.67      1

In [6]:
df.describe()

wingA      wingB
count  20.000000  20.000000
mean   10.397500   8.122500
std     1.370032   1.417195
min     8.360000   5.280000
25%     9.310000   6.747500
50%    10.240000   8.485000
75%    11.215000   9.235000
max    13.670000  10.610000

# Normality of WingA:-

### Criteria 1=0.65 for x1=11.7675 x2=9.0275
### Criteria 2=0.95 for x1=13.1375 x2=7.6575
### Criteria 3=1 for x1=14.507 x2=6.2875

In [7]:
freq_table=pd.crosstab(df['wingB'],'count')  #crosstab is the function in pandas library.
pd.set_option('display.max_rows',None) #To display the full freq table.
freq_table

col_0  count
wingB       
5.28       1
6.38       1
6.57       1
6.61       1
6.68       1
6.77       1
7.20       1
7.57       1
7.95       1
8.44       1
8.53       1
8.89       1
8.92       1
8.95       1
9.23       1
9.25       1
9.29       1
9.30       1
10.03      1
10.61      1

# Normality for wingB:-

#### Criteria 1= 0.65 for x1=9.539 x2=6.7099
#### Criteria 2=1 for x1=5.2878 x2=10.95
#### Criteria 3=1 for x1=3.87 x2=12.37

In [8]:
t_statistic,p_value=ttest_ind(df['wingA'],df['wingB'])
print('tstat %1.3f' % t_statistic)
print('P Value %1.5f' % p_value)

tstat 5.162
P Value 0.00001


#### The value from python is 5.162 and the critical value from the table is 2.101 therefore the alternate hypothesis is accepted,which means there is a significant difference in the performance of students betweeen the two wings.

# How to create a dataframe in python:-

#### If an external file is not provided then we need to create the dataframe directly in python. 

### For creating the dataframe in python for the variable wing A:-

#### data=[1,2,3,4,5.....20] .Using this object data we are trying to create a dataframe df.

#### df=pd.DataFrame(data,columns=['wingA'])

### Find the mean of wingA and cross check with the mean of the imported file.

### Now create a new dataframe with two variables wingA and WingB.and then calculate the mean of the two variables and crosscheck it.

# For wingA

In [9]:
data=[10.7,9.89,11.83,9.04,9.37,11.68,8.36,9.76,13.67,8.96,9.51,10.85,10.57,11.06,8.91,11.79,10.59,9.13,12.37,9.91]
df1=pd.DataFrame(data,columns=['wingA'])

In [10]:
df1.describe()

wingA
count  20.000000
mean   10.397500
std     1.370032
min     8.360000
25%     9.310000
50%    10.240000
75%    11.215000
max    13.670000

# For both wingA and wingB

In [11]:
data1=[[10.7,7.2],[9.89,6.68],[11.83,9.29],[9.04,8.95],[9.37,6.61],[11.68,8.53],[8.36,8.92],[9.76,7.95],[13.67,7.57],[8.96,6.38]
       ,[9.51,8.89],[10.85,10.03],[10.57,9.3],[11.06,5.28],[8.91,9.23],[11.79,9.25],[10.59,8.44],[9.13,6.57],[12.37,10.61],
         [9.91,6.77]] #Inserting the data of both the wings one by one in the subsequent manner.
df2=pd.DataFrame(data1,columns=['wingA','wingB'])

In [12]:
df2.describe() #The data is same as above.

wingA      wingB
count  20.000000  20.000000
mean   10.397500   8.122500
std     1.370032   1.417195
min     8.360000   5.280000
25%     9.310000   6.747500
50%    10.240000   8.485000
75%    11.215000   9.235000
max    13.670000  10.610000

# Paired sample t-test:-

### For doing the paired sample t-test,the student should should remain the same for both the variables.
## The function to be used is 'ttest_rel'.
#### Import the function from scipy.stats.
#### The degree of freedom is (n-1) where n is the sample size.
# Assumption for doing the pair sample t-test:-
### Create a new variable which is the difference of variable 1 and variable 2 and then check for its normality.
### if this new variable has a normal distribution then use pair sample t-test,otherwise use a parametric test wilcoxon .
### Create a dataframe from the data.
### Create a new variable which is the difference of the two variable.
### Check the normality for this new variable.
### Conduct the pair sample t-test and write your interpretation using the statistical table.


In [13]:
data2=[[1,2.83,3.505],[2,3.295,3.43],
[3,2.71,3.67],
[4,2.855,3.355],
[5,2.98,3.985],
[6,3.065,3.63],
[7,3.765,4.57],
[8,3.265,3.7],
[9,3.17,3.66],
[10,2.895,3.25],
[11,2.63,2.85],
[12,2.83,3.34],
[13,2.935,3.63],
[14,3.115,3.675],
[15,2.985,3.475],
[16,3.135,3.605],
[17,2.75,3.25],
[18,3.205,3.54],
[19,3,4.005],
[20,3.035,3.595]]

df2=pd.DataFrame(data2,columns=['sample','Two Days','Seven Days'])



In [14]:
df2.describe()

sample   Two Days  Seven Days
count  20.00000  20.000000   20.000000
mean   10.50000   3.022500    3.586000
std     5.91608   0.253229    0.345458
min     1.00000   2.630000    2.850000
25%     5.75000   2.848750    3.411250
50%    10.50000   2.992500    3.600000
75%    15.25000   3.143750    3.671250
max    20.00000   3.765000    4.570000

In [15]:
df2.head()

sample  Two Days  Seven Days
0       1     2.830       3.505
1       2     3.295       3.430
2       3     2.710       3.670
3       4     2.855       3.355
4       5     2.980       3.985

In [16]:
df2['difference']=df2['Two Days']-df2['Seven Days']
df2

sample  Two Days  Seven Days  difference
0        1     2.830       3.505      -0.675
1        2     3.295       3.430      -0.135
2        3     2.710       3.670      -0.960
3        4     2.855       3.355      -0.500
4        5     2.980       3.985      -1.005
5        6     3.065       3.630      -0.565
6        7     3.765       4.570      -0.805
7        8     3.265       3.700      -0.435
8        9     3.170       3.660      -0.490
9       10     2.895       3.250      -0.355
10      11     2.630       2.850      -0.220
11      12     2.830       3.340      -0.510
12      13     2.935       3.630      -0.695
13      14     3.115       3.675      -0.560
14      15     2.985       3.475      -0.490
15      16     3.135       3.605      -0.470
16      17     2.750       3.250      -0.500
17      18     3.205       3.540      -0.335
18      19     3.000       4.005      -1.005
19      20     3.035       3.595      -0.560

In [17]:
df2.describe()

sample   Two Days  Seven Days  difference
count  20.00000  20.000000   20.000000    20.00000
mean   10.50000   3.022500    3.586000    -0.56350
std     5.91608   0.253229    0.345458     0.23835
min     1.00000   2.630000    2.850000    -1.00500
25%     5.75000   2.848750    3.411250    -0.68000
50%    10.50000   2.992500    3.600000    -0.50500
75%    15.25000   3.143750    3.671250    -0.46125
max    20.00000   3.765000    4.570000    -0.13500

In [18]:
freq_table=pd.crosstab(df2['difference'],'count') 
pd.set_option('display.max_rows',None) 
freq_table

col_0       count
difference       
-1.005          2
-0.960          1
-0.805          1
-0.695          1
-0.675          1
-0.565          1
-0.560          1
-0.560          1
-0.510          1
-0.500          2
-0.490          2
-0.470          1
-0.435          1
-0.355          1
-0.335          1
-0.220          1
-0.135          1

### Normality of difference:-
##### Criteria1=0.75 for x1=-0.80185 x2=-0.32515
##### Criteria2=1 for x1=-0.0868 x2=-1.0402
##### Criteria3=1 for x1=0.15155 x2=1.27855
#### Normal distribution

In [19]:
from scipy.stats import ttest_rel

In [20]:
t_statistics,p_value=ttest_rel(df2['Two Days'],df2['Seven Days'])
t_statistics

-10.572884939003538

In [21]:
print('tstat %1.3f' % t_statistics)
print('P Value %1.5f' % p_value)

tstat -10.573
P Value 0.00000


### The t value obtained from python is 10.57 and the critical value from the statistical table is 2.09 for 19 degrees of freedom.Thereforn the null hypothesis is rejected and the teaching has been effective.

## Use the file hsb2.csv and do the pair sample t-test with the interpretation for the following combinations of variables:- 1. read and write 2. maths and science 3. science and socst

# How to do Mann Whitney U Test:-
### To do the mann Whitney we need a function in python "mannwhitneyu" from scipy.stats Library.
### For doing it:
### stat,p=mannwhitneyu(df['read'],df['write'])
### Find the value of p from the above equation.
## If the p<0.05 (5% significance level) then the null hypothesis is rejected which means there is a significance difference between read and write.
## If the p>=0.05 then the null hypothesis is accepted which means there is no significance difference between read and write.

# Wilcoxon Text:-
## If the difference between the two variables doesn't have a normal distribution then use the wilcoxon test.
### import "wilcoxon" from scipy.stats Library.
### For doing it:-
### stat,p=wilcoxon(df['read'],df[;write'])


# One Sample t-test:-
### Objective is to see whether the p
### Null hypothesis(Ho)=There is no significant difference in the mean of read marks against the benchmark value of 60.
### Alternate hypothesis(Ha)=There is the significant difference in the mean of read marks against the benchmark of 60.
## Therefore if the Null hypothesis is accepted that the performance of the students has met my expections.
### import "ttest_1samp" from scipy.stats library.
### t_statistics,p_value=ttest_1samp(df['read'],the benchmark value) and make the interpretation of the p value from it.

# Format to write Interpretation:-
## 1. Objective
## 2.Null Hypothesis
## 3.Alternate Hypothesis
## 4.Which Test
## 5.Statistical parameter obtained
## 6.Interpretation from the test

## Q1:- To see whether the performance of the students in the science subject has met a benchmark value of 60.
## Q2:-To see is there a significance difference in the math subject between 0 and 1.
## Q3:- To see is there a significance difference between science and socst.
## Q4:- To see is there a significance difference in science subject between student of schtyp 1 and 2
## Q5:- To see is there a significance difference science and socst and the data is coming from different students.

# Objective1:-To see whether the performance of the students in the science subject has met a benchmark value of 60:-
### Null hypothesis(Ho)=There is no significant difference in the mean of science subject against the benchmark value of 60.
### Alternate hypothesis(Ha)=There is the significant difference in the mean of science subject against the benchmark of 60.
## We will do the One Sample t-test in this case.


In [22]:
df3=pd.read_csv('F:\hsb2.csv')

In [23]:
from scipy.stats import ttest_1samp

In [24]:
t_statistic,p_value=ttest_1samp(df3['science'],60)
p_value

3.171144080035124e-24

In [25]:
print('P Value %1.5f' % p_value)

P Value 0.00000


# Interpretation-The null hypothesis is rejected as the value of p value is zero which is less than 0.05.
# Hence we can say that the students has met the benchmark of 60 in the science subject.

# Objective2:-To see is there a significance difference in the math subject between 0 and 1.

### Null Hypothesis:- There is no significant difference in the math subject between 0 and 1.
### Alternate Hypothesis:- There is a significant difference in the math subject between 0 and 1.

## We will do the independent t-test

In [26]:
df3.describe()

id     Gender        race         ses      schtyp        prog  \
count  200.000000  200.00000  200.000000  200.000000  200.000000  200.000000   
mean   100.500000    0.54500    3.430000    2.055000    1.160000    2.025000   
std     57.879185    0.49922    1.039472    0.724291    0.367526    0.690477   
min      1.000000    0.00000    1.000000    1.000000    1.000000    1.000000   
25%     50.750000    0.00000    3.000000    2.000000    1.000000    2.000000   
50%    100.500000    1.00000    4.000000    2.000000    1.000000    2.000000   
75%    150.250000    1.00000    4.000000    3.000000    1.000000    2.250000   
max    200.000000    1.00000    4.000000    3.000000    2.000000    3.000000   

             read       write        math     science       socst  
count  200.000000  200.000000  200.000000  200.000000  200.000000  
mean    52.230000   52.775000   52.645000   51.850000   52.405000  
std     10.252937    9.478586    9.368448    9.900891   10.735793  
min     28.000000   31.000000   33.000000   26.000000   26.000000  
25%     44.000000   45.750000   45.000000   44.000000   46.000000  
50%     50.000000   54.000000   52.000000   53.000000   52.000000  
75%     60.000000   60.000000   59.000000   58.000000   61.000000  
max     76.000000   67.000000   75.000000   74.000000   71.000000

In [27]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   id       200 non-null    int64
 1   Gender   200 non-null    int64
 2   race     200 non-null    int64
 3   ses      200 non-null    int64
 4   schtyp   200 non-null    int64
 5   prog     200 non-null    int64
 6   read     200 non-null    int64
 7   write    200 non-null    int64
 8   math     200 non-null    int64
 9   science  200 non-null    int64
 10  socst    200 non-null    int64
dtypes: int64(11)
memory usage: 17.3 KB


### Check the normality for math.If it is normal then independent sample t-test otherwise mann whitney test.

In [28]:
t_statistic,p_value=ttest_ind(df3['math'][df3['Gender']==0],df3['math'][df3['Gender']==1])
t_statistic

0.41299864929688507

In [29]:
freq_table=pd.crosstab(df3['math'],'count')
pd.set_option('display.max_rows',None) 
freq_table

col_0  count
math        
33         1
35         1
37         1
38         2
39         6
40        10
41         7
42         7
43         7
44         4
45         8
46         8
47         3
48         5
49        10
50         7
51         8
52         6
53         7
54        10
55         5
56         7
57        13
58         6
59         2
60         5
61         7
62         4
63         5
64         5
65         3
66         4
67         2
68         1
69         2
70         1
71         4
72         3
73         1
75         2

## Normality for math:-
### Criteria1=0.66 x1=62 x2=43.27
### Criteria2=0.97 x1=71.376 x2=33.90
### Criteria3=1 x1=24.54 x2=80.74
# The math variable has normal distribution.

## So null hypothesis is accepted; Degree of Freedom:n-1=199 
# Interpretation:- There is no significant difference between 0 and 1 in the math subject. 

# Objective 3:-To see is there a significance difference between science and socst.
## Null Hypothesis:-There is no significant difference between science and socst.
## Alternate Hypothesis:-There is a significant difference between science and socst.
## We will do the Paired sample t test.

In [30]:
df3['difference']=df3['science']-df3['socst']

In [31]:
df3

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0     70       0     4    1       1     1    57     52    41       47     57   
1    121       1     4    2       1     3    68     59    53       63     61   
2     86       0     4    3       1     1    44     33    54       58     31   
3    141       0     4    3       1     3    63     44    47       53     56   
4    172       0     4    2       1     2    47     52    57       53     61   
5    113       0     4    2       1     2    44     52    51       63     61   
6     50       0     3    2       1     1    50     59    42       53     61   
7     11       0     1    2       1     2    34     46    45       39     36   
8     84       0     4    2       1     1    63     57    54       58     51   
9     48       0     3    2       1     2    57     55    52       50     51   
10    75       0     4    2       1     3    60     46    51       53     61   
11    60       0     4    2       1     2    57     65    51       63     61   
12    95       0     4    3       1     2    73     60    71       61     71   
13   104       0     4    3       1     2    54     63    57       55     46   
14    38       0     3    1       1     2    45     57    50       31     56   
15   115       0     4    1       1     1    42     49    43       50     56   
16    76       0     4    3       1     2    47     52    51       50     56   
17   195       0     4    2       2     1    57     57    60       58     56   
18   114       0     4    3       1     2    68     65    62       55     61   
19    85       0     4    2       1     1    55     39    57       53     46   
20   167       0     4    2       1     1    63     49    35       66     41   
21   143       0     4    2       1     3    63     63    75       72     66   
22    41       0     3    2       1     2    50     40    45       55     56   
23    20       0     1    3       1     2    60     52    57       61     61   
24    12       0     1    2       1     3    37     44    45       39     46   
25    53       0     3    2       1     3    34     37    46       39     31   
26   154       0     4    3       1     2    65     65    66       61     66   
27   178       0     4    2       2     3    47     57    57       58     46   
28   196       0     4    3       2     2    44     38    49       39     46   
29    29       0     2    1       1     1    52     44    49       55     41   
30   126       0     4    2       1     1    42     31    57       47     51   
31   103       0     4    3       1     2    76     52    64       64     61   
32   192       0     4    3       2     2    65     67    63       66     71   
33   150       0     4    2       1     3    42     41    57       72     31   
34   199       0     4    3       2     2    52     59    50       61     61   
35   144       0     4    3       1     1    60     65    58       61     66   
36   200       0     4    2       2     2    68     54    75       66     66   
37    80       0     4    3       1     2    65     62    68       66     66   
38    16       0     1    1       1     3    47     31    44       36     36   
39   153       0     4    2       1     3    39     31    40       39     51   
40   176       0     4    2       2     2    47     47    41       42     51   
41   177       0     4    2       2     2    55     59    62       58     51   
42   168       0     4    2       1     2    52     54    57       55     51   
43    40       0     3    1       1     1    42     41    43       50     41   
44    62       0     4    3       1     1    65     65    48       63     66   
45   169       0     4    1       1     1    55     59    63       69     46   
46    49       0     3    3       1     3    50     40    39       49     47   
47   136       0     4    2       1     2    65     59    70       63     51   
48   189       0     4    2       2     2    47     59    63       53     46   
49    

In [32]:
freq_table=pd.crosstab(df3['difference'],'count')
pd.set_option('display.max_rows',None) 
freq_table

col_0       count
difference       
-25             2
-22             1
-20             1
-17             3
-16             5
-14             6
-13             5
-12            12
-11             3
-10             3
-9              1
-8             10
-7              9
-6             10
-5              4
-4              5
-3             10
-2              4
-1             10
 0              9
 1             10
 2             10
 3             11
 4              2
 5              4
 6              3
 7              7
 8              4
 9              6
 10             6
 11             1
 12             5
 14             3
 16             2
 17             1
 18             1
 20             1
 21             1
 22             1
 23             1
 24             2
 25             2
 27             1
 32             1
 41             1

In [33]:
df3.describe()

id     Gender        race         ses      schtyp        prog  \
count  200.000000  200.00000  200.000000  200.000000  200.000000  200.000000   
mean   100.500000    0.54500    3.430000    2.055000    1.160000    2.025000   
std     57.879185    0.49922    1.039472    0.724291    0.367526    0.690477   
min      1.000000    0.00000    1.000000    1.000000    1.000000    1.000000   
25%     50.750000    0.00000    3.000000    2.000000    1.000000    2.000000   
50%    100.500000    1.00000    4.000000    2.000000    1.000000    2.000000   
75%    150.250000    1.00000    4.000000    3.000000    1.000000    2.250000   
max    200.000000    1.00000    4.000000    3.000000    2.000000    3.000000   

             read       write        math     science       socst  difference  
count  200.000000  200.000000  200.000000  200.000000  200.000000  200.000000  
mean    52.230000   52.775000   52.645000   51.850000   52.405000   -0.555000  
std     10.252937    9.478586    9.368448    9.900891   10.735793   10.696215  
min     28.000000   31.000000   33.000000   26.000000   26.000000  -25.000000  
25%     44.000000   45.750000   45.000000   44.000000   46.000000   -8.000000  
50%     50.000000   54.000000   52.000000   53.000000   52.000000   -1.000000  
75%     60.000000   60.000000   59.000000   58.000000   61.000000    5.250000  
max     76.000000   67.000000   75.000000   74.000000   71.000000   41.000000

## Normality of Difference:
#### criteria1=0.69 for x1=10.14 x2=-11.24
#### criteria2=0.935 for x1=20.88 x2=-21.88
#### criteria3=0.995 for x1=31.57 x2=-32.57

### The difference variable has not a normal distribution.So we will do the Wilcoxon Test.

In [34]:
from scipy.stats import wilcoxon

In [35]:
t_statistic,p_value=wilcoxon(df3['science'],df3['socst'])

In [36]:
t_statistic

8044.5

In [37]:
p_value

0.14167018734360068

## The value of p is greater than 0.05 so the null hypothesis is accepted.
# Interpretation:-So there is no significance difference between science and socst.

 # Objective 4:-To see is there a significance difference in science subject between student of schtyp 1 and 2.

In [38]:
df4=pd.read_csv('F:\hsb2.csv')

In [39]:
freq_table=pd.crosstab(df4['science'],'count')
pd.set_option('display.max_rows',None) 
freq_table

col_0    count
science       
26           1
29           1
31           2
33           2
34           5
35           1
36           4
39          13
40           2
42          12
44          11
45           1
46           1
47          11
48           2
49           2
50          21
51           2
53          15
54           3
55          18
56           2
57           1
58          19
59           1
61          14
63          12
64           1
65           1
66           9
67           1
69           6
72           2
74           1

### Normality for Science:-
1. Criteria 1st Area=>0.68  x1=61.75 x2=41.95
2. Criteria 2nd Area=>0.965 x1=71.65 x2=32.05
3. Criteria 3rd Area=>1     x1=22.15 x2=81.55
#### The science variable has a normal Distribution.So we will apply independent t test.

In [40]:
t_statistic,p_value=ttest_ind(df4['science'][df4['schtyp']==1],df4['science'][df4['schtyp']==2])
t_statistic

-0.9113221460646189

### The critical value of the science variable is 1.972(approaximately) which is greater than the calculated t value from the python.So the Null hypothesis is accepted.

## Interpretation: We Can conclude that there is no significance difference between students of schtyp for science variable.

# Objective 5:-To see is there a significance difference science and socst and the data is coming from different students.